src/term.rs

```diff
@@ -183,14 +183,69 @@ impl Default for BindingType {
     }
 }
 
+/// A contract with its associated data.
+#[derive(Debug, PartialEq, Clone)]
+pub struct PendingContract {
+    /// The "pending" contract, can be a function or a record.
+    pub contract: RichTerm,
+    /// The blame label.
+    pub label: Label,
+}
+
+impl PendingContract {
+    pub fn new(contract: RichTerm, label: Label) -> Self {
+        PendingContract { contract, label }
+    }
+}
+
 /// The attributes of an Array.
-#[derive(Debug, Default, Eq, PartialEq, Copy, Clone)]
+#[derive(Debug, Default, PartialEq, Clone)]
 pub struct ArrayAttrs {
     /// A `closurized` array verifies the following conditions:
     ///   - Each element is a generated variable with a unique name (although the same
     ///     variable can occur in several places, it should always refer to the same thunk anyway).
     ///   - The environment of the array's closure only contains those generated variables.
     pub closurized: bool,
+    /// List of lazily-applied contracts.
+    /// These are only observed when data enters or leaves the array.
+    pub pending_contracts: Vec<PendingContract>,
+}
+
+impl ArrayAttrs {
+    /// Create an `ArrayAttrs`.
+    /// By default, the `closurized` flag is set to `false`,
+    /// and `pending_contracts` is empty.
+    pub fn new() -> Self {
+        Default::default()
+    }
+
+    /// Set the `closurized` flag to `true`.
+    pub fn closurized(mut self) -> Self {
+        self.closurized = true;
+        self
+    }
+
+    /// Drop the pending contracts.
+    pub fn contracts_cleared(mut self) -> Self {
+        self.pending_contracts.clear();
+        self
+    }
+
+    /// Extend contracts from an iterator of `PendingContract`.
```

Comment:
```suggestion
    /// Extend contracts from an iterator of `PendingContract`.
    /// De-duplicate equal contracts. Note that current contract
    /// equality testing is very limited, but this may change in the
    /// future
```